src/Std/Data/HashSet/Basic.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+prelude
+import Std.Data.HashMap.Basic
+
+/-!
+# Hash sets
+
+This file develops the type `Std.Data.HashSet` of dependent hash sets.
+
+The operations `map` and `filterMap` on `Std.Data.HashSet` are defined in the module
+`Std.Data.HashSet.AdditionalOperations`.
+
+Lemmas about the operations on `Std.Data.HashSet` are available in the
+module `Std.Data.HashSet.Lemmas`.
+
+See the module `Std.Data.HashSet.Raw` for a variant of this type which is safe to use in
+nested inductive types.
+-/
+
+set_option linter.missingDocs true
+set_option autoImplicit false
+
+universe u v
+
+variable {α : Type u}
+
+namespace Std
+
+/-- Hash sets. -/
+structure HashSet (α : Type u) [BEq α] [Hashable α] where
+  /-- Internal implementation detail of the hash set. -/
+  inner : HashMap α Unit
+
+namespace HashSet
+
+/-- Creates a new empty hash set. The optional parameter `capacity` can be supplied to presize the
+map so that it can hold the given number of elements without reallocating. It is also possible to
+use the empty collection notations `∅` and `{}` to create an empty hash map with the default
+capacity. -/
+@[inline] def empty [BEq α] [Hashable α] (capacity := 8) : HashSet α :=
+  ⟨HashMap.empty capacity⟩
+
+instance [BEq α] [Hashable α] : EmptyCollection (HashSet α) where
+  emptyCollection := empty
+
+/-- Insert the given element into the set. -/
+@[inline] def insert [BEq α] [Hashable α] (m : HashSet α) (a : α) : HashSet α :=
+  ⟨m.inner.insertIfNew a ()⟩
```

Comment:
Shouldn't this unconditionally insert the element?

Here's a silly example:
```
structure Mod5 where
  val : Nat 

instance : BEq Mod5 where
  beq x y := x.val % 5 == y.val % 5

instance : Hashable Mod5 where
  hash x := hash x.val
```
Putting these in a `HashSet Mod5` would mean that I could keep track of counters, but only one for each equivalence class. If I were doing this, then I would expect that inserting a value would update the value for its equivalence class, rather than being a no-op.

This is a silly way to write that code, and `HashMap (Fin 5) Nat` or just a five-tuple would much more clearly show the intention, but I hope it was illustrative. At the very least, I think there should be a prominent warning if the `BEq` will be interpreted as indicating that all elements of the `BEq`'s equivalence class are interchangeable.

Reply:
The operation you want is called `replace` in the Rust hash set and if different from `insert` which behaves like the `insert` I define. The Java hash set does not have this operation at all. I'm not saying that wanting this is unreasonable, but I think that we should keep the `insert` operation as is and add a new operation `replace` in the future if we want it.

Reply:
That seems reasonable to me